### DISCUSSION TITLE
getting SNPs

### DISCUSSION THREAD

**Discussion by laurel-nave-powers** [sr320/course-fish546-2021#130] [Q&A] [answered]
I have my alignment files in .sam format (yay) and now I want to call SNPs. I have been reading in Chapter 11 Working with Alignment Data and it seems I have two potential options - IGV (pg 374-375) or `samtools mpileup` (pg 378). Both seem to require an input of a specific region to look for SNPs in and I'm not sure how I am supposed to get that. 
Please let me know if I am super far off in this approach and which option would be easiest/fastest. Thanks!

**Comment by sr320** [comment]
mpileup is correct


>

**Reply by sr320** [reply]
first you would convert sam to bam... sorting and indexing a la

```
samtools view -bS /Volumes/caviar/wd/2016-12-01/1NF_9A.sam \
| samtools sort -o /Volumes/caviar/wd/2016-12-01/1NF_9A.bam
samtools index /Volumes/caviar/wd/2016-12-01/1NF_9A.bam
```

then mpileup a la 

```
samtools mpileup --VCF --no-BAQ \
--fasta-ref /Users/sr320/git-repos/student-fish546-2016/data/Ostrea_lurida-Scaff-10k.fa \
/Volumes/caviar/wd/2016-12-02/1NF_33A.bam \
-o /Volumes/caviar/wd/2016-12-02/1NF_33A.vcf.gz
```

**Reply by laurel-nave-powers** [reply]
Alright I think I'm having issues installing samtools. I followed the build and install instructions at http://www.htslib.org/download/ 
and this is what my code looks like  https://github.com/fish546-2021/Laurel-genes/blob/main/code/SNP.ipynb

**Reply by laurel-nave-powers** [reply]
Ok I re-added it to my PATH and it can find samtools now, however the code doesn't seem to be working... is it not liking the flags I'm using? https://github.com/fish546-2021/Laurel-genes/blob/main/code/SNP.ipynb

**Reply by kubu4** [reply]
Change this:

```bash
!/Applications/bioinfo/samtools1.11/bin/samtools \
samtools view -bS ../analyses/HIPP-DR.sam \
| samtools sort -o ../analyses/HIPP-DR.bam \
samtools index ../analyses/HIPP-DR.bam
```

to this:

```bash
!samtools view -bS ../analyses/HIPP-DR.sam \
| samtools sort -o ../analyses/HIPP-DR.bam \
samtools index ../analyses/HIPP-DR.bam
```

**Reply by laurel-nave-powers** [reply]
Now it's saying failed to write to file (wrong size) 
https://github.com/fish546-2021/Laurel-genes/blob/main/code/SNP.ipynb

**Reply by kubu4** [reply]
My bad, change to this:

```
!samtools view -bS ../analyses/HIPP-DR.sam \
| samtools sort -o ../analyses/HIPP-DR.bam -
samtools index ../analyses/HIPP-DR.bam
```

The _critical_ change here is the addition of the hyphen at the end of the `samtools sort` command. That tells `samtools` that the input file is actually being passed in from `stdout` of the previous command (`samtools view` in this case).

I also deleted the line continuation backslash, as it's not needed and ends up adding a bit of confusion.

**Reply by laurel-nave-powers** [reply]
Hmmm it gave me the same error. 
Also, it kept giving me errors when I have it as three lines, so I made it into two lines.

**Reply by kubu4** [reply]
Did that fix it, though (making into two lines)?

Also, I just realized it's because of differences in your cell formatting and mine. I always put `%%bash` as the first line of my Jupyter cells. That way, it will run multiple lines of bash code, without the need for line continuation backslashes.

Without the `%%bash`, the Jupyter cell will only execute a single line of bash code.

Sorry for the oversight on that.

**Reply by laurel-nave-powers** [reply]
I've been googling but not finding anything super helpful. 
Would it have anything to do with the version of samtools I have downloaded (1.11)? That's popped up in some of the googling but I'm not sure it's applicable to me.

**Reply by laurel-nave-powers** [reply]
No it didn't fix it, it still gives me the same error :( the notebook is updated https://github.com/fish546-2021/Laurel-genes/blob/main/code/SNP.ipynb

**Reply by laurel-nave-powers** [reply]
Is the fact that it starts with "usage: samtools sort" mean the problem is in the sort part of the code?

**Reply by kubu4** [reply]
Replace the contents of your cell with the following:

```
%%bash
# Convert SAM to sorted BAM
samtools view -bS ../analyses/HIPP-DR.sam \
| samtools sort -o ../analyses/HIPP-DR.sorted.bam -

# Create BAM index file
samtools index ../analyses/HIPP-DR.sorted.bam
```

Let me know how it goes.

P.S. Changes to filenames that I made won't make a difference to success/fail and aren't actually needed. I just added them for clarity; personal preference. Descriptive filenames helped me catch a _major_ screw-up that I made on a project of mine last week. Not sure I would have caught the problem without having the descriptive filenames.

**Reply by laurel-nave-powers** [reply]
Been running for a bit now, no errors yet so that's a good sign, I will let you know what happens when it finishes!

**Reply by laurel-nave-powers** [reply]
I think it worked? It gave me the .sorted.bam file. It also gave me this message highlighted in red (does red mean error?):  "[bam_sort_core] merging from 13 files and 1 in-memory blocks... "
But since it gave me the .sorted.bam file I think it's ok. Thanks for your help!    
https://github.com/fish546-2021/Laurel-genes/blob/main/code/SNP.ipynb

**Comment by laurel-nave-powers** [comment]
Error with the next step using mpileup: https://github.com/fish546-2021/Laurel-genes/blob/main/code/SNP.ipynb 
Seems to be an issue with the reference genome? Does it not like that it's not unzipped?

**Reply by kubu4** [reply]
Correct.

**Reply by laurel-nave-powers** [reply]
Shoot, when I try to unzip it it says no space left on device, the file must be too big?

**Reply by kubu4** [reply]
>  it says no space left on device, the file must be too big?

Unfortunately, yes.

Find some (big) things you can delete and try again?

**Reply by laurel-nave-powers** [reply]
I'm deleting as much as I can but it's still not enough space. The things taking up the most space are the sam and bam files that I need to use for this so I don't think I can delete them. 
Not sure what to do if I can't get it to unzip. 
Is there a way to tell how much space it needs to be unzipped? Its 693 MB in .gz format

**Reply by kubu4** [reply]
You can delete SAM file.

**Reply by laurel-nave-powers** [reply]
Awesome, it all worked! It gave me the .vcf.gz file.  https://github.com/fish546-2021/Laurel-genes/blob/main/code/SNP.ipynb 
I think I'm a little lost conceptually. What is an "la"? 
I'm trying to line up what I have done with what's explained in the book. On page 380 it says the gzipped vcf file is intermediate genotype and variant data. I think my next step to use `bcftools call` ? 
Basically I'm trying to figure out where I'm at and what I should do next if my end goal is just to call SNPs.

**Reply by kubu4** [reply]
> What is an "la"?

Where is this from? Not sure what you're referring to.

**Reply by laurel-nave-powers** [reply]
See Stevens comment on this discussion post, he refers to indexing and mpileup on an la.

**Reply by kubu4** [reply]
Gotcha. He's saying "a la" as in "e.g." or "like so".

**Reply by laurel-nave-powers** [reply]
Hahahaha ahhh I see. I wondered why I wasn't finding that language with any of the samtools text. Thank you. 
I am going to use `bcftools call` next unless I'm missing something conceptually there.

**Comment by laurel-nave-powers** [comment]
I used `bcftools call` and `zgrep` to call and see variant sites. https://github.com/fish546-2021/Laurel-genes/blob/main/code/SNP.ipynb 
Please let me know if there is more I need to do with this.

**Comment by sr320** [comment, marked as answer]
I think you are there!!!

Nice Job!

**Reply by laurel-nave-powers** [reply]
Yay! Thank you for all your help
